src/common/secret/src/vault_client.rs

```diff
@@ -0,0 +1,423 @@
+    pub fn from_protobuf(vault_backend: &secret::SecretHashicorpVaultBackend) -> Result<Self> {
+        let auth = match vault_backend.auth.as_ref() {
+            Some(secret::secret_hashicorp_vault_backend::Auth::TokenAuth(token_auth)) => {
+                HashiCorpVaultAuth::Token {
+                    auth_token: token_auth.token.clone(),
+                }
+            }
+            Some(secret::secret_hashicorp_vault_backend::Auth::ApproleAuth(approle_auth)) => {
+                HashiCorpVaultAuth::AppRole {
+                    auth_role_id: approle_auth.role_id.clone(),
+                    auth_secret_id: approle_auth.secret_id.clone(),
+                }
+            }
+            None => {
+                return Err(anyhow::anyhow!(
+                    "No auth method specified for Vault backend"
+                ));
+            }
+        };
+
+        Ok(HashiCorpVaultConfig {
+            addr: vault_backend.addr.clone(),
+            path: vault_backend.path.clone(),
+            field: vault_backend.field.clone(),
+            auth,
+            tls_skip_verify: vault_backend.tls_skip_verify,
+        })
+    }
+
+    /// Convert `HashiCorpVaultConfig` to protobuf `SecretHashicorpVaultBackend`
+    pub fn to_protobuf(&self) -> secret::SecretHashicorpVaultBackend {
+        let auth = match &self.auth {
+            HashiCorpVaultAuth::Token { auth_token } => Some(
+                secret::secret_hashicorp_vault_backend::Auth::TokenAuth(secret::VaultTokenAuth {
+                    token: auth_token.clone(),
+                }),
+            ),
+            HashiCorpVaultAuth::AppRole {
+                auth_role_id,
+                auth_secret_id,
+            } => Some(secret::secret_hashicorp_vault_backend::Auth::ApproleAuth(
+                secret::VaultAppRoleAuth {
+                    role_id: auth_role_id.clone(),
+                    secret_id: auth_secret_id.clone(),
+                },
+            )),
+        };
+
+        secret::SecretHashicorpVaultBackend {
+            addr: self.addr.clone(),
+            path: self.path.clone(),
+            field: self.field.clone(),
+            auth,
+            tls_skip_verify: self.tls_skip_verify,
+        }
+    }
+}
+
+impl HashiCorpVaultClient {
+    pub fn new(config: HashiCorpVaultConfig) -> Result<Self> {
+        let mut client_builder = Client::builder();
+
+        if config.tls_skip_verify {
+            client_builder = client_builder.danger_accept_invalid_certs(true);
+        }
+
+        let client = client_builder
+            .build()
+            .context("Failed to create HTTP client")?;
+
+        Ok(Self { client, config })
+    }
+
+    pub async fn get_secret(&self) -> Result<Vec<u8>> {
+        // Try to get secret, with retry logic for token invalidation
+        let mut force_refresh_token = false;
+
+        // Retry loop for handling token invalidation
+        for retry_count in 0..1 {
```

Comment:
This looks a little odd `0..1` means we only run this loop once, at `0`. `1` is excluded from the iteration.
This means the loop is not necessary.

Reply:
Good catch, the loop here serves as a retry. If cached approle auth token expires, we can re-request one and update cache.